jinahub/indexers/query/vector/AnnoyIndexer/__init__.py

```diff
@@ -17,43 +20,82 @@ class AnnoyIndexer(BaseNumpyIndexer):
         Annoy package dependency is only required at the query time.
     """
 
-    def __init__(self, metric: str = 'euclidean', n_trees: int = 10, search_k: int = -1, *args, **kwargs):
+    ANNOY_INDEX_FILE_NAME = 'annoy.ann'
+    ANNOY_INDEX_MAPPING_NAME = (
+        'map.json'  # this map stores int id and docid, annoy only allows add int as id.
+    )
+
+    def __init__(
+        self,
+        top_k: int = 10,
+        num_dim: int = 768,
+        metric: str = 'euclidean',
+        num_trees: int = 10,
+        traverse_path: list = ['r'],
+        save_on_close: bool = True,
+        **kwargs,
+    ):
         """
         Initialize an AnnoyIndexer
 
+        :param top_k: get tok k vectors
+        :param num_dim: dimension of the vector
         :param metric: Metric can be "angular", "euclidean", "manhattan", "hamming", or "dot"
-        :param n_trees: builds a forest of n_trees trees. More trees gives higher precision when querying.
-        :param search_k: At query time annoy will inspect up to search_k nodes which defaults to
-            n_trees * k if not provided (set to -1)
+        :param num_trees: builds a forest of n_trees trees. More trees gives higher precision when querying.
+        :param traverse_path: traverse path on docs, e.g. ['r'], ['c']
         :param args:
         :param kwargs:
         """
-        super().__init__(*args, compress_level=0, **kwargs)
+        super().__init__(**kwargs)
+        self.top_k = top_k
         self.metric = metric
-        self.n_trees = n_trees
-        self.search_k = search_k
+        self.num_dim = num_dim
+        self.num_trees = num_trees
+        self.id_docid_map = {}
+        self.index_base_dir = f'{kwargs["metas"]["workspace"]}/annoy/'
+        self.save_on_close = save_on_close
+        self.index_path = self.index_base_dir + self.ANNOY_INDEX_FILE_NAME
+        self.index_map_path = self.index_base_dir + self.ANNOY_INDEX_MAPPING_NAME
+        self.indexer = AnnoyIndex(self.num_dim, self.metric)
+        self.traverse_path = traverse_path
+        if os.path.exists(self.index_path) and os.path.exists(self.index_map_path):
```

Comment:
Not sure I understand this. Annoy is a query-type Indexer. It should not store its state to disk to be re-read. Instead it should take a `dump_path` as an argument and then call the import_vectors and import_ids from that dump. 

See similar pattern in NumpyIndexer https://github.com/jina-ai/indexers/blob/69279e78f2e437e50a23270e915abc3d7c4a7499/jinahub/indexers/query/vector/NumpyIndexer/__init__.py#L21

Reply:
when and where should we dump the indexed data?

Reply:
The Query indexer does not dump. That is the job of the DBMS-type Indexers. You can create a dummy dump for your test and just import from that. Look at the tests for NpIndexer https://github.com/jina-ai/indexers/blob/9adcead316a3ed92daf6dce546726cf1d2d5cf4a/jinahub/indexers/query/vector/NumpyIndexer/tests/test_query.py#L6-L5

Reply:
yeah this pattern is great